### PR TITLE
[instantiation linking] Create and import WASMTableInstance

### PR DESCRIPTION
**Pull request by lum1n0us**
Based on https://github.com/bytecodealliance/wasm-micro-runtime/pull/3893

### CODE REVIEW

File: core/iwasm/interpreter/wasm_runtime.c [hunk 1]
```diff
@@ -713,29 +725,58 @@ wasm_destroy_memory(WASMMemoryInstance *memory)
...
+    WASMModule *module = module_inst->module;
+    /* only imported tables */
+    for (uint32 i = 0; i < module->import_table_count; i++) {
+        WASMTableInstance *table = module_inst->tables[i];
+        if (!table) {
+            continue;
+        }
+
+        table_elem_type_t *table_elems =
+            wasm_locate_table_elems(module, table, i);
```
Comment (wenyongh) [inline core/iwasm/interpreter/wasm_runtime.c:746]: Same as aot_runtime.c, should we check whether `table_elems` is NULL, and free it only when it is not NULL?

File: core/iwasm/aot/aot_loader.c [hunk 2]
```diff
@@ -1451,6 +1473,13 @@ load_table_list(const uint8 **p_buf, const uint8 *buf_end, AOTModule *module,
         if (wasm_is_type_multi_byte_type(table->table_type.elem_type)) {
             read_uint8(buf, buf_end, ref_type.ref_ht_common.nullable);
         }
+#else
+        {
+            /* TBC: placeholder */
+            uint8 placeholder;
+            read_uint8(buf, buf_end, placeholder);
```
[the remark this answers is not in the document]
Comment (lum1n0us) [reply core/iwasm/aot/aot_loader.c:1480]: `aot_emit_table_info()` always `EMIT_U8(something);`. https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/compilation/aot_emit_aot_file.c#L1962C1-L1973C10. 

It's more like a bug-fix to me.

Btw, is it better if I patch it in main branch then merge it back?

File: core/iwasm/aot/aot_loader.c [hunk 3]
```diff
@@ -1364,8 +1372,8 @@ load_init_expr(const uint8 **p_buf, const uint8 *buf_end, AOTModule *module,
 
 static bool
 load_import_table_list(const uint8 **p_buf, const uint8 *buf_end,
-                       AOTModule *module, char *error_buf,
-                       uint32 error_buf_size)
+                       AOTModule *module, bool is_load_from_file_buf,
+                       char *error_buf, uint32 error_buf_size)
```
[the remark this answers is not in the document]
Comment (lum1n0us) [reply core/iwasm/aot/aot_loader.c:1376]:
```c
        if (module->package_version >= 4) {
            read_string(buf, buf_end, import_table->module_name); // -> load_string(...)
            read_string(buf, buf_end, import_table->table_name);
        }
```

File: core/iwasm/aot/aot_loader.c [hunk 4]
```diff
@@ -1582,14 +1611,15 @@ load_table_init_data_list(const uint8 **p_buf, const uint8 *buf_end,
 
 static bool
 load_table_info(const uint8 **p_buf, const uint8 *buf_end, AOTModule *module,
-                char *error_buf, uint32 error_buf_size)
+                bool is_load_from_file_buf, char *error_buf,
+                uint32 error_buf_size)
```
[the remark this answers is not in the document]
Comment (lum1n0us) [reply core/iwasm/aot/aot_loader.c:1615]: `read_string(...)`

File: core/iwasm/aot/aot_loader.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (wenyongh) [reply core/iwasm/aot/aot_loader.c:1376]: got it, thanks

File: core/iwasm/aot/aot_loader.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (wenyongh) [reply core/iwasm/aot/aot_loader.c:1480]: Got it, it seems that the below `read_uint32(buf, buf_end, table->table_type.init_size)` will adjust the ptr to 4-byte aligned address, so this byte can be bumped over and it should work when GC isn't enabled. But patching it in main branch makes the code more friendly.

File: core/iwasm/aot/aot_runtime.c [hunk 5]
```diff
@@ -5608,3 +5733,42 @@ aot_disinherit_imports(AOTModule *module, WASMExternInstance *imports,
...
+#if WASM_ENABLE_GC != 0
+    table->elem_ref_type.elem_ref_type = type->elem_ref_type;
+#endif
+
+    /* Set all elements to -1 or NULL_REF to mark them as uninitialized
+     * elements */
+#if WASM_ENABLE_GC == 0
+    memset(table->elems, 0xff, sizeof(table_elem_type_t) * table->max_size);
+#else
+    memset(table->elems, 0x00, sizeof(table_elem_type_t) * table->max_size);
```
[the remark this answers is not in the document]
Comment (lum1n0us) [reply core/iwasm/aot/aot_runtime.c:5761]: I prefer to use clear assignment to display different initial values for various features.

File: core/iwasm/fast-jit/jit_frontend.c [hunk 6]
```diff
@@ -706,6 +706,20 @@ get_table_elems_reg(JitFrame *frame, uint32 tbl_idx)
...
                  NEW_CONST(PTR, offset));
     }
+
+    if (tbl_idx < cc->cur_wasm_module->import_table_count) {
+        /*
+         * If the table is imported, the table_elems is a pointer to the
+         * imported table's table_elems.
+         * need to load the real one from the
+         */
+        /* table_elems = *(void**)table_elems */
```
[the remark this answers is not in the document]
Comment (lum1n0us) [reply core/iwasm/fast-jit/jit_frontend.c:716]: Fast-JIT typically does not support import tables; each table is local. Therefore, there is no necessity to use a macro.

File: core/iwasm/aot/aot_runtime.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (wenyongh) [reply core/iwasm/aot/aot_runtime.c:5761]: OK

File: core/iwasm/fast-jit/jit_frontend.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (wenyongh) [reply core/iwasm/fast-jit/jit_frontend.c:716]: Got it, thanks.

File: core/iwasm/compilation/aot_emit_aot_file.c [hunk 7]
```diff
@@ -370,71 +368,58 @@ get_table_init_data_list_size(AOTCompContext *comp_ctx,
...
-     * ------------------------------
-     */
-    uint32 size = 0, i;
+    /* import_table_count */
+    uint32 size = (uint32)sizeof(uint32);
 
-    size = (uint32)sizeof(uint32);
-    for (i = 0; i < comp_data->import_table_count; i++) {
+    for (uint32 i = 0; i < comp_data->import_table_count; i++) {
+        size = align_uint(size, 2);
```
[the remark this answers is not in the document]
Comment (wenyongh) [reply core/iwasm/compilation/aot_emit_aot_file.c:391]: OK, had better remove L391

[the remark this answers is not in the document]
Comment (wenyongh) [reply core/iwasm/compilation/aot_emit_aot_file.c:391]: Sorry, found that this line is in a for loop and the end of loop is calculating the size of string table_name, so the previous `size = align_uint(size, 4);` cannot guarantee that here it is 4-byte aligned.

File: core/iwasm/compilation/aot_emit_aot_file.c [hunk 8]
```diff
@@ -370,71 +368,58 @@ get_table_init_data_list_size(AOTCompContext *comp_ctx,
...
+        /*
+         *   u8 elem_type + u8 flags
+         *   + u8 possible_grow + u8 elem_ref_type.nullable
+         * ------------------------------ =u32
+         *   + u32 init_size + u32 max_size
+         *   + u32 elem_ref_type.heap_type
+         */
         size += sizeof(uint32) * 3;
```
[the remark this answers is not in the document]
Comment (lum1n0us) [reply core/iwasm/compilation/aot_emit_aot_file.c:399]:
``` c
// aot_emit_import_table_info()
        EMIT_U8(import_table->table_type.elem_type);
        EMIT_U8(import_table->table_type.flags);
        EMIT_U8(import_table->table_type.possible_grow);
#if WASM_ENABLE_GC != 0
        if (comp_ctx->enable_gc && import_table->table_type.elem_ref_type) {
            EMIT_U8(
                import_table->table_type.elem_ref_type->ref_ht_common.nullable);
        }
        else
#endif
        {
            /* emit one placeholder to keep the same size */
            EMIT_U8(0);
        }  
        // --------------------- U32
        EMIT_U32(import_table->table_type.init_size);
        EMIT_U32(import_table->table_type.max_size);
        // --------------------- U32 * 3
```

File: core/iwasm/compilation/aot_emit_aot_file.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (lum1n0us) [reply core/iwasm/compilation/aot_emit_aot_file.c:391]: Got it.

File: core/iwasm/compilation/aot_emit_aot_file.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (wenyongh) [reply core/iwasm/compilation/aot_emit_aot_file.c:399]: got it, thanks.

File: core/iwasm/aot/aot_loader.c [hunk 9]
```diff
@@ -1391,6 +1399,13 @@ load_import_table_list(const uint8 **p_buf, const uint8 *buf_end,
         if (wasm_is_type_multi_byte_type(import_table->table_type.elem_type)) {
             read_uint8(buf, buf_end, ref_type.ref_ht_common.nullable);
         }
+#else
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c:1402]: had better add an `else` branch when WASM_ENABLE_GC is enabled, how about:
```C
#if WASM_ENABLE_GC != 0
        if (wasm_is_type_multi_byte_type(import_table->table_type.elem_type)) {
            read_uint8(buf, buf_end, ref_type.ref_ht_common.nullable);
        }
        else
#endif
        {
            ...
        }
```

File: core/iwasm/aot/aot_loader.c [hunk 10]
```diff
@@ -1451,6 +1473,13 @@ load_table_list(const uint8 **p_buf, const uint8 *buf_end, AOTModule *module,
         if (wasm_is_type_multi_byte_type(table->table_type.elem_type)) {
             read_uint8(buf, buf_end, ref_type.ref_ht_common.nullable);
         }
+#else
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c:1476]: same as above, how about:
```C
#if WASM_ENABLE_GC != 0
        if (wasm_is_type_multi_byte_type(table->table_type.elem_type)) {
            read_uint8(buf, buf_end, ref_type.ref_ht_common.nullable);
        }
        else
#endif
        {
            ...
        }
```

File: core/iwasm/aot/aot_loader.c [hunk 11]
```diff
@@ -1409,6 +1424,13 @@ load_import_table_list(const uint8 **p_buf, const uint8 *buf_end,
             }
         }
 #endif
+
+        /* TBC: sync up with AOT_CURRENT_VERSION in config.h */
+        /* v3 doesn't emit names */
+        if (module->package_version >= 4) {
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c:1430]: Seems that the import table in the older aot file isn't used at all or developer doesn't generate the import table? Maybe we can remove the line?

**Review by wenyongh — APPROVED** [review]
LGTM with minor comments.

File: core/iwasm/aot/aot_loader.c (hunk 11 above)
Comment (lum1n0us) [reply core/iwasm/aot/aot_loader.c:1430]: I'm not certain. I'll add a TODO note here, maintain the 'if' control, and run a full compatible test later.

File: core/iwasm/aot/aot_loader.c [hunk 12]
```diff
@@ -1392,6 +1400,13 @@ load_import_table_list(const uint8 **p_buf, const uint8 *buf_end,
             read_uint8(buf, buf_end, ref_type.ref_ht_common.nullable);
         }
 #endif
+        {
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c:1403]: should add `else` before `#endif`

File: core/iwasm/aot/aot_loader.c [hunk 13]
```diff
@@ -1452,6 +1479,13 @@ load_table_list(const uint8 **p_buf, const uint8 *buf_end, AOTModule *module,
...
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c:1482]: should add else before #endif

File: core/iwasm/aot/aot_loader.c (hunk 12 above)
Comment (lum1n0us) [reply core/iwasm/aot/aot_loader.c:1403]: sync up with `aot_emit_table_info()`

```c
#if WASM_ENABLE_GC != 0
        if (comp_ctx->enable_gc
            && comp_data->import_tables[i].table_type.elem_ref_type) {
            EMIT_U8(comp_data->import_tables[i]
                        .table_type.elem_ref_type->ref_ht_common.nullable);
        }
        else
#endif
        {
            /* emit one placeholder to keep the same size */
            EMIT_U8(0);
        }

```

File: core/iwasm/aot/aot_loader.c (hunk 13 above)
Comment (lum1n0us) [reply core/iwasm/aot/aot_loader.c:1482]: sync up with `aot_emit_table_info()`

```c
#if WASM_ENABLE_GC != 0
        if (comp_ctx->enable_gc
            && comp_data->tables[i].table_type.elem_ref_type) {
            EMIT_U8(comp_data->tables[i]
                        .table_type.elem_ref_type->ref_ht_common.nullable);
        }
        else
#endif
        {
            /* emit one placeholder to keep the same size */
            EMIT_U8(0);
        }
```

File: core/iwasm/aot/aot_loader.c (hunk 12 above)
Comment (wenyongh) [reply core/iwasm/aot/aot_loader.c:1403]: ok, so you will add `else` before `#endif`?
```
#if WASM_ENABLE_GC != 0
        if (wasm_is_type_multi_byte_type(import_table->table_type.elem_type)) {
            read_uint8(buf, buf_end, ref_type.ref_ht_common.nullable);
        }
        else
#endif
        {
            ...
        }
```

Comment (lum1n0us) [reply core/iwasm/aot/aot_loader.c:1403]: 😨 oops

**Review by wenyongh — APPROVED** [review]
LGTM

**Pull request merged**